clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -253,6 +284,58 @@ public void forceClose() {
         initiateClose();
     }
 
+    private void sendInitPidRequest(Node node) {
+        String nodeId = node.idString();
+        InitPidRequest.Builder builder = new InitPidRequest.Builder(null);
+        RequestCompletionHandler callback = new RequestCompletionHandler() {
+            public void onComplete(ClientResponse response) {
+                handleInitPidResponse(response);
+            }
+        };
+        ClientRequest request = client.newClientRequest(nodeId, builder, time.milliseconds(), true, callback);
+        client.send(request, time.milliseconds());
+    }
+
+    private Node getReadyNode(long remainingTimeMs) {
```

Comment:
Take a look at `NetworkClientBlockingOps.blockingReady()` since this is super tricky to get right. It definitely needs to be a loop because `poll()` may return for reasons other than the event we are looking for. The call to `connectionFailed` also needs to be done after the call to `poll()` since `ready()` will reset the state.